src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyProgressService.java

```diff
@@ -217,26 +204,187 @@ public CompetencyProgress updateCompetencyProgress(Long competencyId, User user)
+            CompetencyProgress competencyProgress) {
+        double numberOfLearningObjects = lectureUnits.size() + exerciseInfos.size();
+        if (numberOfLearningObjects == 0) {
+            // If nothing is linked to the competency, the competency is considered completed
+            competencyProgress.setProgress(100.0);
+        }
+
+        double achievedPoints = exerciseInfos.stream().mapToDouble(info -> info.lastPoints() != null ? info.lastPoints() : 0).sum();
+        double maxPoints = exerciseInfos.stream().mapToDouble(CompetencyExerciseMasteryCalculationDTO::maxPoints).sum();
+        double exerciseProgress = maxPoints > 0 ? achievedPoints / maxPoints * 100 : 0;
+
+        double lectureProgress = 100.0 * numberOfCompletedLectureUnits / lectureUnits.size();
+
+        double progress = exerciseInfos.size() / numberOfLearningObjects * exerciseProgress + lectureUnits.size() / numberOfLearningObjects * lectureProgress;
+        // Bonus points can lead to a progress > 100%
+        progress = Math.clamp(Math.round(progress), 0, 100);
+        competencyProgress.setProgress(progress);
+    }
+
+    /**
+     * Calculate the confidence score for the given user in a competency based on the exercises linked to the competency.
      *
-     * @param learningObjects A list of all learning objects linked to a specific competency
-     * @param user            The user for which the progress should be calculated
-     * @return The percentage of completed learning objects by the user
+     * @param exerciseInfos      The information about the exercises linked to the competency
+     * @param competencyProgress The progress entity to update
      */
-    private double calculateProgress(@NotNull Set<LearningObject> learningObjects, @NotNull User user) {
-        return learningObjects.stream().map(learningObject -> learningObjectService.isCompletedByUser(learningObject, user)).mapToInt(completed -> completed ? 100 : 0).average()
-                .orElse(0.);
+    private void calculateConfidence(Set<CompetencyExerciseMasteryCalculationDTO> exerciseInfos, CompetencyProgress competencyProgress) {
+        Set<CompetencyExerciseMasteryCalculationDTO> participantScoreInfos = exerciseInfos.stream()
+                .filter(info -> info.lastScore() != null && info.lastPoints() != null && info.lastModifiedDate() != null).collect(Collectors.toSet());
+
+        double recencyConfidenceHeuristic = calculateRecencyConfidenceHeuristic(participantScoreInfos);
+        double difficultyConfidenceHeuristic = calculateDifficultyConfidenceHeuristic(participantScoreInfos, exerciseInfos);
+        double quickSolveConfidenceHeuristic = calculateQuickSolveConfidenceHeuristic(participantScoreInfos);
+
+        double confidence = 1 + recencyConfidenceHeuristic + difficultyConfidenceHeuristic + quickSolveConfidenceHeuristic;
```

Comment:
We could use a variable e.g. `MIN_CONFIDENCE` for the `1` to improve the readability - I do not have a strong opinion here

Reply:
Added comment above this line, so the developer directly sees that the base factor is one and does not have to look up what `MIN_CONFIDENCE` is

Reply:
Actually added comment and constant